### ISSUE TITLE
h5 file content stays in memory after close() and is overwritten

### ISSUE THREAD

**Issue by nicoloabrate** [h5py/h5py#2069]
* Windows 10, CentOS 7
 * Python 3.8.12 (on both)
 * Anaconda (on both)
 * h5py version: 3.6.0 on Windows and 2.10.0 on CentOS 7
 * HDF5 version: 1.10.6 on Windows, 1.8.12 on CentOS 7

I am using h5py to read a specific dataset from a file produced by an external code, written in Fortran. This dataset is a simple array of instants in time, from 0 up to 3. In my python script I need to manipulate this array, for example I pass it to "np.argwhere" to look for some specific time index. When I do this and I read again the dataset, the first value of the dataset changes.
Here you can find a minimum working example. 

````
import h5py as h5
import numpy as np
from copy import deepcopy as cp
# I read the file
fh5 = h5.File("C:\\Users\\Desktop\\output.h5", "r")
times = fh5['distributions']['timeDistr'][()]
fh5.close()
# the file seems to be closed
print(fh5)

print(times[0:3])

times[0] = 10
print(times[0:3])

# here I mimic some manipulations of the original dataset
ttt = []
for t in times:
    ttt.append(t)
ttt = cp(np.asarray(ttt))

tmp2 = list([ttt+42]) # this affects the dataset
tmp3 = list(np.asarray(tmp2)+10) # this does not affect the dataset

# I read again the dataset
fh52 = h5.File("C:\\Users\\Desktop\\output.h5", "r")
newtimes = fh52['distributions']['timeDistr'][()]
fh52.close()

print(newtimes[0:3])
````
The file "output.h5" is attached here as a zip file.
[output.zip](https://github.com/h5py/h5py/files/8335475/output.zip)

In this example I manipulate the dataset after dumping it into an array. When I read again the file _ex novo_, the dataset is changed! The first value is the only to be affected, it becomes 52 in this case, which is the results of the operations ``times[0] = 10`` and ``tmp2 = list([ttt+42])``. This is akward, especially because ``list([ttt+42])`` is assigned to another variable which has nothing to do with the dataset.

How can it be possible? Did I something extremely stupid? I never saw something like this, it seems completely crazy to me!

Thank you so much for your help, 
Nicolò

**Comment by tacaswell** [comment]
I can reproduce this.......it is wild........

**Comment by tacaswell** [comment]
There is something about _this_ file and _this_ dataset.  Trying to re-create this with a smaller data set yielded a working file and if you make a copy of the dataset into a copy of the file also works again.

```python
import h5py as h5
import numpy as np

if False:
    with h5.File("/tmp/output.h5", "a") as fout:
        with h5.File("/tmp/output_orign.h5", "r") as fh5:
            del fout["distributions/timeDistr"]
            fout.create_dataset(
                "distributions/timeDistr",
                data=fh5["distributions/timeDistr"][()],
                maxshape=(None,),
                chunks=(1,),
            )

N = 1
fname = "/tmp/output.h5"
# fname = '/tmp/test.h5'
# I read the file
with h5.File(fname, "r") as fh5:
    times = fh5["distributions/timeDistr"][()]

# the file seems to be closed
print(fh5)

print(times[:N])
ttt = np.array(times)
ttt + 1
with h5.File(fname, "r") as fh52:
    newtimes = fh52["distributions/timeDistr"][()]
print(newtimes[:N])

arrays = []
for j in range(5):
    print("about to loop")
    with h5.File(fname, "r") as fin:
        r = fin["distributions/timeDistr"][()]
    g = np.array(r, copy=True)
    print(f"share memory: {np.shares_memory(r, g)}")

    print(f"{j}: pre add {g[:N]}")
    g + j
    print(f"{j}: post add {g[:N]}")
    arrays.append(g[0])


```


```
<Closed HDF5 file>
[0.]
[1.]
about to loop
share memory: False
0: pre add [6.90146063e-310]
0: post add [6.90146063e-310]
about to loop
share memory: False
1: pre add [6.90146063e-310]
1: post add [6.90146063e-310]
about to loop
share memory: False
2: pre add [1.]
2: post add [1.]
about to loop
share memory: False
3: pre add [3.]
3: post add [3.]
about to loop
share memory: False
4: pre add [6.]
4: post add [6.]

```


h5repack leaves us with a file that has this issue.

**Comment by tacaswell** [comment]
Weird things that I noticed:

  1. the first value is tiny, I think sub-normal
  2. the chunksize is (1, ) which is a bit pathological
  3. no compression

**Comment by tacaswell** [comment]
And I'm testing with very new versions of things

```
[] print(h5py.version.info)
Summary of the h5py configuration
---------------------------------

h5py    3.6.0
HDF5    1.12.1
Python  3.11.0a6+ (heads/main:29624e769c, Mar 21 2022, 23:11:58) [GCC 11.2.0]
sys.platform    linux
sys.maxsize     9223372036854775807
numpy   1.23.0.dev0+902.gfc9044863
cython (built with) 3.0.0a10
numpy (built against) 1.23.0.dev0+902.gfc9044863
HDF5 (built against) 1.12.1

```

**Comment by tacaswell** [comment]
```
import h5py as h5
import numpy as np

N = 1
fname = "/tmp/output.h5"

with h5.File(fname, "r", rdcc_nslots=1) as fin:
    for k in range(5):
        r = fin["distributions/timeDistr"][:]
        g = np.array(r, copy=True)
        print(fin["/distributions/timeDistr"][:][:N], g[:N], r[:N])
        g + 1
        
```

gives 

```
[6.92868218e-310] [0.] [0.]
[0.] [1.] [1.]
[1.] [2.] [2.]
[2.] [3.] [3.]
[3.] [4.] [4.]

```

**Comment by nicoloabrate** [comment]
Thanks to the observations done by @tacaswell, I contacted the developer of the Fortran code that produces this output file and a bug was found and fixed: the dataset for time was written skipping the first time (the initial condition), implying that some typical Fortran-garbage was put in the first chunk. This explains the presence of 6.92868218e-310.

Nevertheless, as user of h5py, I think that the behaviour of the file in Python is too weird and dangerous, as this "bug" sneaks behind apparently trivial operations, so it should be prevented in some way, if possible!

**Comment by tacaswell** [comment]
Ok, This I propose the following theory:
 - libh5 is able to tell there is no data in the first chunk (which is reasonable) so does not write to it
 - when we read from h5 we make a numpy array of the right size and then pass that to libh5 and let it fill the array with data.  Very likely we use `np.empty` to not waste time setting memory we are about to let libhdf5 set.
 - numpy keeps a pool of allocated arrays (which I imagine speeds up temporaries because there is no need to allocate ram, just grab one that is big enough, use the head of it and then give it back to the pool).
 - first pass through you get garbage (which seems to be the garbage that `h5ls` also gets which is ... odd but carrying on)
 - doing something that creates a temporary array will pull something from the pool, mutate it, and then drop it back into the pool with out cleaning it (assumed to be next user of the memory block's job)
 - the next time we read we grab that temporary and write all but the first element
 - the old result from the first element leaks through because no one ever cleared it

I think this story explains all of the things we were seeing.

The question is how can we detect when we hand an array to libh5 that does not get fully written to and either raise on put in the fill value?

I'm not sure if this story is completely correct, but I feel a lot better having an at least plausible to me story and expect this will not keep me up again tonight....

**Comment by nicoloabrate** [comment]
Your theory totally satisfies me. To my knowledge, this most likely explains the strange behaviour observed for such cases!! Unfortunately, I am not expert enough to help answering to the question regarding the handling of arrays that do not get fully written...
In my experience, the best option, provided that we are able to detect "missing" data and garbage replacement, would be to print a warning and to yield a dataset filled with NaN in the missing locations. In such a case, it would be up to the user to decide if this is fine or if there is a bug in the h5 writing process, as it was in my case.

**Comment by takluyver** [comment]
> when we read from h5 we make a numpy array of the right size and then pass that to libh5 and let it fill the array with data. Very likely we use np.empty to not waste time setting memory we are about to let libhdf5 set.

I think this is essentially correct. The code in the fast path (which I'm guessing this will use) calls `PyArray_SimpleNew`:

https://github.com/h5py/h5py/blob/3051c76afa73ad0c72d2c7c6724fe3617d2fefd8/h5py/_selector.pyx#L332

The [docs for which](https://numpy.org/doc/stable/reference/c-api/array.html#c.PyArray_SimpleNew) say that it uses unitialized memory. The slow path calls `ndarray()`:

https://github.com/h5py/h5py/blob/3051c76afa73ad0c72d2c7c6724fe3617d2fefd8/h5py/_hl/dataset.py#L792

The docs don't seem to specify for that, but [the implementation calls](https://github.com/numpy/numpy/blob/beb5dedf78b96183906ec9e979ef83bb040b7b78/numpy/core/src/multiarray/arrayobject.c#L1624-L1628) calls `PyArray_NewFromDescr_int`, passing 0 for the `zeroed` parameter, so that's also unitialized.

Looking back, it seems we've actually had an issue about this for years - #455. A fix was attempted in #456, but abandoned because "Does not work for object types". It looks like `np.zeros()` is meant to handle object dtype OK - maybe this has changed in the intervening time? Although it does fill with 0 (as an object) rather than `None`. It's not clear what exactly went wrong back then.

**Comment by takluyver** [comment]
I've revived the possible fix in #2076. I haven't figured out exactly what the issue with it was last time, though.
